README
NAME

    Git::Background - use Git commands with Future

VERSION

    Version 0.007_02

SYNOPSIS

        use Git::Background 0.008;

        my $git = Git::Background->new($dir);
        my $future = $git->run('status', '-s');
        my @status = $future->stdout;

        my $future = Git::Background->run('status', '-s', { dir => $dir });
        my @status = $future->stdout;

USAGE

 new( [DIR], [ARGS] )

    Creates and returns a new Git::Background object. If you specify the
    dir positional argument, or use the dir argument in the args hash the
    directory is passed as cwd option to Proc::Background causing it to
    change into that directory before running the Git command.

    Both, the dir positional argument and the args hash are optional. An
    exception is thrown if you specify both.

        my $git = Git::Background->new;
        my $git = Git::Background->new($dir);
        my $git = Git::Background->new( { dir => $dir, fatal => 0 } );
        my $git = Git::Background->new( $dir, { fatal => 0 } );

    new either returns a valid Git::Background object or throws an
    exception.

    The following options can be passed in the args hash to new. They are
    used as defaults for calls to run.

    Current API available since 0.001.

  dir

    This will be passed as cwd argument to Proc::Background whenever you
    call run. If you don't specify a dir all Git commands are executed in
    whatever the current working directory is when you call run.

  fatal

    Enabled by default. The fatal option controls if "await" in
    Git::Background::Future returns a failed Future when Git returns a
    non-zero return code.

    Please not that "await" in Git::Background::Future always returns a
    failed Future if Git returns 128 (fatal Git error) or 129 (Git usage
    error) regardless of fatal. And a failed Future is returned if another
    error happens, e.g. if the output from Git cannot be read.

  git

    The Git command used to run. This defaults to git and lets
    Proc::Background work its magic to find the binary on your platform.

    This can be either a string,

        my $git = Git::Background->new( { git => '/opt/git/bin/git' } );

    or an array ref.

        my $git = Git::Background->new({
            git => [ qw( /usr/bin/sudo -u nobody git ) ],
        });

 run( @CMD, [ARGS] )

    This runs the specified Git command in the background by passing it on
    to Git::Background. The last argument can be an argument hash that
    takes the same arguments as new.

        my $git = Git::Background->new($dir);
        my $future = $git->run('status', '-s', { fatal => 0 } );

        if ( $future->await->is_failed ) {
            say q{Unable to run 'git status -s'};
        }
        else {
            my @status = $future->stdout;
        }

    The call returns a Git::Background::Future and the Git command runs in
    its own process. All output produced by Git is redirected to a
    File::Temp temporary file.

    Proc::Background is run with autoterminate set, which will kill the Git
    process if the future is destroyed.

    Since version 0.004 run croaks if it gets called in void context.

    Current API available since 0.004.

 version( [ARGS] )

    Returns the version of the used Git binary or undef if no Git command
    was found. This call uses the same, optional, argument hash as run. The
    call is wrapped in an eval which ensures that this method never throws
    an error and can be used to check if a Git is available.

        my $version = Git::Background->version;
        if ( !defined $version ) {
            say "No Git binary found.";
        }
        else {
            say "You have Git version $version";
        }

    Current API available since 0.001.

EXAMPLES

 Example 1 Clone a repository

    Cloning a repository is a bit special as it's the only Git command that
    cannot be run in a workspace and the target directory must not exist.

    There are two ways to use a Git::Background object without the
    workspace directory:

        my $future = Git::Background->run('clone', $url, $dir);
        $future->get;

        # later, use a new object for working with the cloned repository
        my $git = Git::Background->new($dir);
        my $future = $git->run('status', '-s');
        my @stdout = $future->stdout;

    Alternatively you can overwrite the directory for the call to clone:

        my $git = Git::Background->new($dir);
        my $future = $git->run('clone', $url, $dir, { dir => undef });
        $future->get;

        # then use the same object for working with the cloned repository
        my $future = $git->run('status', '-s');
        my @stdout = $future->stdout;

SEE ALSO

    Git::Repository, Git::Wrapper, Future, Git::Version::Compare

SUPPORT

 Bugs / Feature Requests

    Please report any bugs or feature requests through the issue tracker at
    https://github.com/skirmess/Git-Background/issues. You will be notified
    automatically of any progress on your issue.

 Source Code

    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    https://github.com/skirmess/Git-Background

      git clone https://github.com/skirmess/Git-Background.git

AUTHOR

    Sven Kirmess <sven.kirmess@kzone.ch>



Changes
Revision history for Git-Background

{{$NEXT}}

0.008  2023-08-17 12:03:02 UTC

  - No changes from 0.007_02.

0.007_02  2023-08-16 12:44:24 UTC

  - get() now returns 2 additional values, a Path::Tiny object for the
    files containing stdout and stderr from git.
  - added 2 new methods, path_stderr, and path_stdout.
  - Remove dependency on Test2::Plugin::UTF8.

0.007_01  2023-08-05 16:02:13 UTC

  - The git used for the tests now always outputs UTF-8, as the real
    Git does.
  - Added a new test to test the UTF-8 handling.

0.007  2023-08-05 12:30:25 UTC

  - use binmode($fh, ':encoding(UTF-8)') on the file handles before reading
    the output from git.
  - Bump required Perl to 5.10, which is required by Future anyway.

0.006  2022-11-05 13:16:26 UTC

  - Switch license to the ISC license
  - Bumped dependency to Future 0.49 and use the new set_udata/udata methods
  - Remove test dependency on Test::Fatal

0.005  2022-04-03 16:05:30 UTC

  - Use Git::Background::Future->fail instead of Future->fail when
    Proc::Background fails. Otherwise mthods like stdout() wouldn't be
    available on these errors but on others, which would be confusing.

0.004  2022-01-08 21:19:34 UTC

  - run() throws an error if it is run in void context

0.003  2022-01-01 22:55:58 UTC

  - Bump dependency on Future to 0.40, which introduced Future::Exception

0.002  2022-01-01 19:40:33 UTC

  - API change to use Future

0.001  2021-12-20 20:01:35 UTC

  - Initial release


Makefile.PL
use 5.010;
use strict;
use warnings;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
    ABSTRACT           => 'use Git commands with Future',
    AUTHOR             => 'Sven Kirmess <sven.kirmess@kzone.ch>',
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    DISTNAME         => 'Git-Background',
    LICENSE          => 'open_source',
    MIN_PERL_VERSION => '5.010',
    NAME             => 'Git::Background',
    PREREQ_PM        => {
        Carp               => 0,
        Future             => '0.49',
        'Path::Tiny'       => '0.125',
        'Proc::Background' => '1.30',
        'Scalar::Util'     => 0,
        parent             => 0,
        strict             => 0,
        warnings           => 0,
    },
    TEST_REQUIRES => {
        Cwd                     => 0,
        Encode                  => 0,
        Exporter                => 0,
        'File::Basename'        => 0,
        'File::Path'            => '2.07',
        'File::Spec'            => 0,
        'File::Spec::Functions' => 0,
        'Test::More'            => '0.88',
        lib                     => 0,
        overload                => 0,
    },
    VERSION => '0.007_02',
    test    => {
        TESTS => 't/*.t',
    },
);

my %FallbackPrereqs = (
    Carp                    => 0,
    Cwd                     => 0,
    Encode                  => 0,
    Exporter                => 0,
    'File::Basename'        => 0,
    'File::Path'            => '2.07',
    'File::Spec'            => 0,
    'File::Spec::Functions' => 0,
    Future                  => '0.49',
    'Path::Tiny'            => '0.125',
    'Proc::Background'      => '1.30',
    'Scalar::Util'          => 0,
    'Test::More'            => '0.88',
    lib                     => 0,
    overload                => 0,
    parent                  => 0,
    strict                  => 0,
    warnings                => 0,
);

if ( $ENV{AUTOMATED_TESTING} || $ENV{EXTENDED_TESTING} ) {
    $WriteMakefileArgs{test}{TESTS} .= ' xt/*.t';
    _add_extended_test_requirements();
}

if ( !eval { ExtUtils::MakeMaker->VERSION('6.63_03') } ) {
    delete $WriteMakefileArgs{TEST_REQUIRES};
    delete $WriteMakefileArgs{BUILD_REQUIRES};
    $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

if ( !eval { ExtUtils::MakeMaker->VERSION(6.52) } ) {
    delete $WriteMakefileArgs{CONFIGURE_REQUIRES};
}

WriteMakefile(%WriteMakefileArgs);

sub test_requires {
    my ( $module, $version_or_range ) = @_;
    $WriteMakefileArgs{TEST_REQUIRES}{$module} = $FallbackPrereqs{$module} = $version_or_range;
    return;
}

sub _add_extended_test_requirements {
    test_requires( 'Test::MockModule', '0.14' );
    return;
}

# vim: ts=4 sts=4 sw=4 et: syntax=perl


lib/Git/Background/Future.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2021-2023 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.010;
use strict;
use warnings;

package Git::Background::Future;

our $VERSION = '0.007_02';

use Future 0.49;
use parent 'Future';

sub new {
    my ( $class, $run ) = @_;

    my $self = $class->SUPER::new;
    $self->set_udata( '_run', $run );
    return $self;
}

sub await {
    my ($self) = @_;

    my $run = $self->udata('_run');

    return $self if !defined $run;

    $self->set_udata( '_run', undef );

    my $e;
    my $ok;
    {
        local $@;    ## no critic (Variables::RequireInitializationForLocalVars)
        $ok = eval {

            # prevent the "(in cleanup) process is already terminated" message
            $run->{_proc}->autoterminate(0);

            # wait for the process to finish
            $run->{_proc}->wait;

            1;
        };

        if ( !$ok ) {

            # The Future->fail exception must be true
            $e = qq{$@} || 'Failed to wait on Git process with Proc::Background';
        }
    }
    return $self->fail( $e, 'Proc::Background' ) if !$ok;

    my $stdout_file = $run->{_stdout};
    my $stderr_file = $run->{_stderr};
    my @stdout;
    my @stderr;
    {
        local $@;    ## no critic (Variables::RequireInitializationForLocalVars)
        $ok = eval {
            $e      = 'Cannot read stdout';
            @stdout = $stdout_file->lines_utf8( { chomp => 1 } );

            $e      = 'Cannot read stderr';
            @stderr = $stderr_file->lines_utf8( { chomp => 1 } );

            1;
        };

        if ( !$ok ) {
            if ( defined $@ && $@ ne q{} ) {
                $e .= ": $@";
            }
        }
    }
    return $self->fail( $e, 'Path::Tiny' ) if !$ok;

    # get exit code and signal from git process
    my $exit_code = $run->{_proc}->exit_code;

    my @result = (
        \@stdout,
        \@stderr,
        $exit_code,
        $stdout_file,
        $stderr_file,
    );

    # git died by a signal
    return $self->fail( 'Git was terminated by a signal', 'Proc::Background', @result ) if $run->{_proc}->exit_signal;

    if (
        # fatal error
        ( $exit_code == 128 )

        # usage error
        || ( $exit_code == 129 )

        # non-zero return code
        || ( $exit_code && $run->{_fatal} )
      )
    {
        my $message = join "\n", @stderr;
        if ( !length $message ) {
            $message = "git exited with fatal exit code $exit_code but had no output to stderr";
        }

        # $run goes out of scope and the file handles and the proc object are freed
        return $self->fail( $message, 'git', @result );
    }

    # $run goes out of scope and the file handles and the proc object are freed
    return $self->done(@result);
}

sub exit_code {
    my ($self) = @_;
    return ( $self->get )[2];
}

sub is_done {
    my ($self) = @_;
    $self->_await_if_git_is_done;
    return $self->SUPER::is_done;
}

sub is_failed {
    my ($self) = @_;
    $self->_await_if_git_is_done;
    return $self->SUPER::is_failed;
}

sub is_ready {
    my ($self) = @_;
    $self->_await_if_git_is_done;
    return $self->SUPER::is_ready;
}

sub path_stderr {
    my ($self) = @_;
    return ( $self->get )[4];
}

sub path_stdout {
    my ($self) = @_;
    return ( $self->get )[3];
}

sub state {    ## no critic (Subroutines::ProhibitBuiltinHomonyms)
    my ($self) = @_;
    $self->_await_if_git_is_done;
    return $self->SUPER::state;
}

sub stderr {
    my ($self) = @_;
    return @{ ( $self->get )[1] };
}

sub stdout {
    my ($self) = @_;
    return @{ ( $self->get )[0] };
}

sub _await_if_git_is_done {
    my ($self) = @_;

    my $run = $self->udata('_run');
    if ( defined $run && !$run->{_proc}->alive ) {
        $self->await;
    }

    return;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Git::Background::Future - use L<Future> with L<Git::Background>

=head1 VERSION

Version 0.007_02

=head1 SYNOPSIS

    use Git::Background 0.008;
    my $future = Git::Background->run(qw(status -s));

    my @stdout = $future->stdout;

    my ($stdout_ref, $stderr_ref, $exit_code, $stdout_path, $stderr_path) =
        $future->get;

=head1 DESCRIPTION

This is a subclass of L<Future>. Please read the excellent documentation of
C<Future> to see what you can do with this module, this man page only
describes the changes to C<Future> specific to L<Git::Background>.

=head2 UTF-8

The default is to read the output from Git on its stdout and stderr as UTF-8.

The strings returned by the C<get>, C<stderr>, and C<stdout> methods can
therefore contain wide characters. When you write this data to a file handle,
you must ensure that the destination also uses a suitable encoding. This is
necessary to correctly handle any wide characters in the data. You can do this
by setting the encoding of the destination filehandle, e.g.:

    binmode(STDOUT, ':encoding(UTF-8)');

=head1 USAGE

=head2 new( RUN )

New C<Git::Background::Future> objects should be constructed by using the
C<run> method of L<Git::Background>.

Current API available since 0.002.

=head2 await

Blocks and waits until the Git process finishes.

Returns the invocant future itself, so it is useful for chaining.

This method is called by C<get> or C<failure>.

See L<Future/await> for more information.

Current API available since 0.002.

=head2 exit_code

Calls C<get>, then returns the exit code of the Git process.

Because this command calls C<get>, the same exceptions can be thrown.

Current API available since 0.002.

=head2 failure

Waits for the running Git process to finish by calling C<await>. Returns
undef if the future finished successfully, otherwise it returns a list with

    $message, $category, @details

For the C<$category> C<git>, C<@details> is a list of C<stdout_ref>,
C<stderr_ref>, C<exit_code>, C<stdout_path>, and C<stderr_path>. See
C<get> for a description of these values.

Current API available since 0.008.

=head2 get

Waits for the running Git process to finish by calling C<await>. Throws a
L<Future::Exception> if the C<Future> didn't finish successfully. Returns
a list of  C<stdout_ref>, C<stderr_ref>, C<exit_code>, C<stdout_path>,
and C<stderr_path>.

    my $git = Git::Background->new($dir);
    my $future = $git->run('status', '-s');
    # waits for 'git status -s' to finish
    my ($stdout_ref, $stderr_ref, $rc) = $future->get;

=head3 stdout_ref

An array reference containing the stdout from git, split into lines
and chomped.

=head3 stderr_ref

An array reference containing the stderr from git, split into lines
and chomped.

=head3 exit_code

The exit code from git.

=head3 stdout_path

A L<Path::Tiny> object of a file containing the stdout from Git. This
can be used to read the data with a different binmode.

=head3 stderr_path

A L<Path::Tiny> object of a file containing the stderr from Git.

Current API available since 0.008.

=head2 is_done

L<Future/is_done>

Current API available since 0.002.

=head2 is_failed

L<Future/is_failed>

Current API available since 0.002.

=head2 is_ready

L<Future/is_ready>

Current API available since 0.002.

=head2 path_stderr

Calls C<get>, then returns the L<Path::Tiny> object for the file
containing the stderr from Git.

Because this command calls C<get>, the same exceptions can be thrown.

Current API available since 0.008.

=head2 path_stdout

Calls C<get>, then returns the L<Path::Tiny> object for the file
containing the stdout from Git.

Because this command calls C<get>, the same exceptions can be thrown.

Current API available since 0.008.

=head2 state

L<Future/state>

Current API available since 0.002.

=head2 stderr

Calls C<get>, then returns all the lines written by the Git command to
stderr.

Because this command calls C<get>, the same exceptions can be thrown.

Note: C<get> returns all the output lines as array reference, C<stderr>
returns a list.

Current API available since 0.002.

=head2 stdout

Calls C<get>, then returns all the lines written by the Git command to
stdout.

Because this command calls C<get>, the same exceptions can be thrown.

Note: C<get> returns all the output lines as array reference, C<stdout>
returns a list.

Current API available since 0.002.

=head1 SEE ALSO

L<Git::Background>, L<Future>, L<Path::Tiny>

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/skirmess/Git-Background/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software. The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/skirmess/Git-Background>

  git clone https://github.com/skirmess/Git-Background.git

=head1 AUTHOR

Sven Kirmess <sven.kirmess@kzone.ch>

=cut


lib/Git/Background/Exception.pm
package Git::Background::Exception;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.001';

use overload (
    q("")    => '_stringify',
    bool     => sub () { return 1 },
    fallback => 1,
);

sub new {
    my ( $class, $args_ref ) = @_;

    my $self = {
        _exit_code => $args_ref->{exit_code},
        _stderr    => $args_ref->{stderr},
        _stdout    => $args_ref->{stdout},
    };
    bless $self, $class;

    return $self;
}

sub exit_code {
    my ($self) = @_;

    return $self->{_exit_code};
}

sub stderr {
    my ($self) = @_;

    my $stderr = $self->{_stderr};
    if ( !defined $stderr ) {
        $stderr = q{};
    }

    my @stderr = split /\n/xsm, $stderr;
    return @stderr;
}

sub stdout {
    my ($self) = @_;

    my $stdout = $self->{_stdout};
    if ( !defined $stdout ) {
        $stdout = q{};
    }

    my @stdout = split /\n/xsm, $stdout;
    return @stdout;
}

sub _stringify {
    my ($self) = @_;

    my $stderr = join "\n", $self->stderr;
    return $stderr if length $stderr;

    my $exit_code = $self->exit_code;
    return 'git exited with a fatal exit code but had no output to stderr' if !defined $exit_code;
    return "git exited with fatal exit code $exit_code but had no output to stderr";
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Git::Background::Exception - exception class for Git::Background

=head1 VERSION

Version 0.001

=head1 SYNOPSIS

A new error object may be generated and thrown as follows:

    die Git::Background::Exception->new({
        stdout => $stdout,
        stderr => $stderr,
        exit_code => $exit_code,
    });


=head1 DESCRIPTION

This object is thrown by Git::Background if the Git process returned a
non-zero return code.

The object stringifies to either the output printed to STDERR by Git or, if
none was printed, a generic error message containing the exit code of Git.

=head1 USAGE

=head2 new( ARGS )

Constructor for this class.

=head3 exit_code (required)

The exit code returned by Git.

=head3 stderr (required)

A string containing all the output printed to STDERR by Git.

=head3 stdout (required)

A string containing all the output printed to STDOUT by Git.

=head2 exit_code

Returns the exit code returned by Git. This is the reason the exception was
thrown.

=head2 stderr

Returns a list of all the output lines printed to standard error by Git.

    my @stderr = $e->stderr;

=head2 stdout

Returns a list of all the output lines printed to standard output by Git.

    my @stdout = $e->stdout;

=head1 EXAMPLES

=head2 Example 1 Catch exception with Feature::Compat::Try

    use 5.014;
    use strict;
    use warnings;

    use lib qw(.. .. lib);

    use Feature::Compat::Try;
    use Scalar::Util qw(blessed);
    use Git::Background;

    my $git = Git::Background->new;
    try {
        $git->run('--invalid-arg')->get;
    }
    catch ($e) {
        die $e if !blessed $e || !$e->isa('Git::Background::Exception');

        my $exit_code = $e->exit_code;
        my $stderr    = join "\n", $e->stderr;
        warn "Git exited with exit code $exit_code\n$stderr\n";
    }

=head2 Example 2 Catch exception with eval

    use 5.006;
    use strict;
    use warnings;

    use lib qw(.. .. lib);

    use Scalar::Util qw(blessed);
    use Git::Background;

    my $git = Git::Background->new;

    if ( !eval { $git->run('--invalid-arg')->get; 1; } ) {
        my $e = $@;
        die $e if !blessed $e || !$e->isa('Git::Background::Exception');

        my $exit_code = $e->exit_code;
        my $stderr    = join "\n", $e->stderr;
        warn "Git exited with exit code $exit_code\n$stderr\n";
    }

=head1 SEE ALSO

L<Git::Wrapper::Exception>

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/skirmess/Git-Background/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software. The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/skirmess/Git-Background>

  git clone https://github.com/skirmess/Git-Background.git

=head1 AUTHOR

Sven Kirmess <sven.kirmess@kzone.ch>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2021 by Sven Kirmess.

This is free software, licensed under:

  The (two-clause) FreeBSD License

=cut

# vim: ts=4 sts=4 sw=4 et: syntax=perl


lib/Git/Background.pm
# vim: ts=4 sts=4 sw=4 et: syntax=perl
#
# Copyright (c) 2021-2023 Sven Kirmess
#
# Permission to use, copy, modify, and distribute this software for any
# purpose with or without fee is hereby granted, provided that the above
# copyright notice and this permission notice appear in all copies.
#
# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.

use 5.010;
use strict;
use warnings;

package Git::Background;

our $VERSION = '0.007_02';

use Carp ();
use Path::Tiny 0.125 ();
use Proc::Background 1.30;
use Scalar::Util ();

use Git::Background::Future;

# Git::Background->new;
# Git::Background->new($dir);
# Git::Background->new( { %options } );
# Git::Background->new( $dir, { %options } );
# options:
#   - dir
#   - fatal     (default 1)
sub new {
    my $class = shift;

  NEW: {
        my $self;

        last NEW if @_ > 2;

        my $dir;

        if (
            @_
            && (
                # first argument is a scalar
                !defined Scalar::Util::reftype( $_[0] )

                # or object
                || defined Scalar::Util::blessed( $_[0] )
            )
          )
        {
            my $arg = shift @_;

            # stringify objects (e.g. Path::Tiny)
            $dir = "$arg";
        }

        last NEW if @_ > 1;

        if (@_) {
            last NEW if !defined Scalar::Util::reftype( $_[0] ) || Scalar::Util::reftype( $_[0] ) ne 'HASH';

            # first/remaining argument is a hash ref
            my $args = shift @_;
            $self = $class->_process_args($args);
        }
        else {
            $self = $class->_process_args;
        }

        if ( defined $dir ) {
            Carp::croak 'Cannot specify dir as positional argument and in argument hash' if exists $self->{_dir};
            $self->{_dir} = $dir;
        }

        bless $self, $class;
        return $self;
    }

    # unknown args
    Carp::croak 'usage: new( [DIR], [ARGS] )';
}

# Git::Background->run( @cmd );
# Git::Background->run( @cmd, { %options } );
sub run {
    my ( $self, @cmd ) = @_;

    Carp::croak 'Cannot use run() in void context. (The git process would immediately get killed.)' if !defined wantarray;    ## no critic (Community::Wantarray)

    my $config;
    if ( @cmd && defined Scalar::Util::reftype( $cmd[-1] ) && Scalar::Util::reftype( $cmd[-1] ) eq 'HASH' ) {
        my $args = pop @cmd;
        $config = $self->_process_args($args);
    }
    else {
        $config = $self->_process_args;
    }

    my $stdout;
    my $stdout_fh;
    my $stderr;
    my $stderr_fh;
    my $e;
    my $ok;
    {
        local $@;    ## no critic (Variables::RequireInitializationForLocalVars)
        $ok = eval {
            $e         = 'Cannot create temporary file for stdout';
            $stdout    = Path::Tiny->tempfile;
            $e         = 'Cannot obtain file handle for stdout temp file';
            $stdout_fh = $stdout->filehandle('>');
            $e         = 'Cannot create temporary file for stdout';
            $stderr    = Path::Tiny->tempfile;
            $e         = 'Cannot obtain file handle for stderr temp file';
            $stderr_fh = $stderr->filehandle('>');
            1;
        };

        if ( !$ok ) {
            if ( defined $@ && $@ ne q{} ) {
                $e .= ": $@";
            }
        }
    }
    Carp::croak $e if !$ok;

    # Proc::Background
    my $proc_args = {
        stdin         => undef,
        stdout        => $stdout_fh,
        stderr        => $stderr_fh,
        command       => [ @{ $config->{_git} }, @cmd ],
        autodie       => 1,
        autoterminate => 1,
        ( defined $config->{_dir} ? ( cwd => $config->{_dir} ) : () ),
    };

    my $proc;
    {
        local @_;    ## no critic (Variables::RequireInitializationForLocalVars)
        local $Carp::Internal{ (__PACKAGE__) } = 1;

        $proc = eval { Proc::Background->new($proc_args); };

        if ( !defined $proc ) {

            # The Future->fail exception must be true
            $e = qq{$@} || 'Failed to run Git with Proc::Background';
        }
    }
    return Git::Background::Future->fail( $e, 'Proc::Background' ) if !defined $proc;

    return Git::Background::Future->new(
        {
            _fatal  => $config->{_fatal},
            _proc   => $proc,
            _stdout => $stdout,
            _stderr => $stderr,
        },
    );
}

sub version {
    my ( $self, $args ) = @_;

    my @cmd = qw(--version);
    if ( defined $args ) {
        Carp::croak 'usage: Git::Background->version([ARGS])' if !defined Scalar::Util::reftype($args) || Scalar::Util::reftype($args) ne 'HASH';

        push @cmd, $args;
    }

    my $version = eval {
        for my $line ( $self->run(@cmd)->stdout ) {
            if ( $line =~ s{ \A git \s version \s }{}xsm ) {
                return $line;
            }
        }

        return;
    };

    return $version;
}

sub _process_args {
    my ( $self, $args ) = @_;

    if ( !defined Scalar::Util::blessed($self) ) {
        $self = {};
    }

    my %args_keys = map { $_ => 1 } keys %{$args};
    my %config    = (
        _fatal => !!1,
        _git   => ['git'],
    );

    # dir
    if ( exists $args->{dir} ) {

        # stringify objects (e.g. Path::Tiny)
        $config{_dir} = "$args->{dir}";
        delete $args_keys{dir};
    }
    elsif ( exists $self->{_dir} ) {
        $config{_dir} = $self->{_dir};
    }

    # fatal
    if ( exists $args->{fatal} ) {
        $config{_fatal} = !!$args->{fatal};
        delete $args_keys{fatal};
    }
    elsif ( exists $self->{_fatal} ) {
        $config{_fatal} = $self->{_fatal};
    }

    # git
    if ( exists $args->{git} ) {
        my $git = $args->{git};
        $config{_git} = [
            ( defined Scalar::Util::reftype($git) && Scalar::Util::reftype($git) eq 'ARRAY' )
            ? @{ $args->{git} }
            : $git,
        ];
        delete $args_keys{git};
    }
    elsif ( exists $self->{_git} ) {
        $config{_git} = [ @{ $self->{_git} } ];
    }

    #
    my @unknown = sort keys %args_keys;
    Carp::croak 'Unknown argument' . ( @unknown > 1 ? 's' : q{} ) . q{: '} . join( q{', '}, sort @unknown ) . q{'} if @unknown;

    return \%config;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Git::Background - use Git commands with L<Future>

=head1 VERSION

Version 0.007_02

=head1 SYNOPSIS

    use Git::Background 0.008;

    my $git = Git::Background->new($dir);
    my $future = $git->run('status', '-s');
    my @status = $future->stdout;

    my $future = Git::Background->run('status', '-s', { dir => $dir });
    my @status = $future->stdout;

=head1 USAGE

=head2 new( [DIR], [ARGS] )

Creates and returns a new C<Git::Background> object. If you specify the
C<dir> positional argument, or use the C<dir> argument in the args hash
the directory is passed as C<cwd> option to L<Proc::Background> causing it
to change into that directory before running the Git command.

Both, the C<dir> positional argument and the args hash are optional. An
exception is thrown if you specify both.

    my $git = Git::Background->new;
    my $git = Git::Background->new($dir);
    my $git = Git::Background->new( { dir => $dir, fatal => 0 } );
    my $git = Git::Background->new( $dir, { fatal => 0 } );

C<new> either returns a valid C<Git::Background> object or throws an
exception.

The following options can be passed in the args hash to new. They are used
as defaults for calls to C<run>.

Current API available since 0.001.

=head3 dir

This will be passed as C<cwd> argument to L<Proc::Background> whenever you
call C<run>. If you don't specify a C<dir> all Git commands are executed in
whatever the current working directory is when you call C<run>.

=head3 fatal

Enabled by default. The C<fatal> option controls if
L<Git::Background::Future/await> returns a failed C<Future> when Git returns a
non-zero return code.

Please not that L<Git::Background::Future/await> always returns a failed
C<Future> if Git returns 128 (fatal Git error) or 129 (Git usage error)
regardless of C<fatal>. And a failed C<Future> is returned if another error
happens, e.g. if the output from Git cannot be read.

=head3 git

The Git command used to run. This defaults to C<git> and lets
L<Proc::Background> work its magic to find the binary on your platform.

This can be either a string,

    my $git = Git::Background->new( { git => '/opt/git/bin/git' } );

or an array ref.

    my $git = Git::Background->new({
        git => [ qw( /usr/bin/sudo -u nobody git ) ],
    });

=head2 run( @CMD, [ARGS] )

This runs the specified Git command in the background by passing it on to
C<Git::Background>. The last argument can be an argument hash that takes the
same arguments as C<new>.

    my $git = Git::Background->new($dir);
    my $future = $git->run('status', '-s', { fatal => 0 } );

    if ( $future->await->is_failed ) {
        say q{Unable to run 'git status -s'};
    }
    else {
        my @status = $future->stdout;
    }

The call returns a L<Git::Background::Future> and the Git command runs in its
own process. All output produced by Git is redirected to a L<File::Temp>
temporary file.

C<Proc::Background> is run with C<autoterminate> set, which will kill the
Git process if the future is destroyed.

Since version 0.004 C<run> C<croaks> if it gets called in void context.

Current API available since 0.004.

=head2 version( [ARGS] )

Returns the version of the used Git binary or undef if no Git command was
found. This call uses the same, optional, argument hash as C<run>. The call
is wrapped in an eval which ensures that this method never throws an error
and can be used to check if a Git is available.

    my $version = Git::Background->version;
    if ( !defined $version ) {
        say "No Git binary found.";
    }
    else {
        say "You have Git version $version";
    }

Current API available since 0.001.

=head1 EXAMPLES

=head2 Example 1 Clone a repository

Cloning a repository is a bit special as it's the only Git command that
cannot be run in a workspace and the target directory must not exist.

There are two ways to use a C<Git::Background> object without the workspace
directory:

    my $future = Git::Background->run('clone', $url, $dir);
    $future->get;

    # later, use a new object for working with the cloned repository
    my $git = Git::Background->new($dir);
    my $future = $git->run('status', '-s');
    my @stdout = $future->stdout;

Alternatively you can overwrite the directory for the call to clone:

    my $git = Git::Background->new($dir);
    my $future = $git->run('clone', $url, $dir, { dir => undef });
    $future->get;

    # then use the same object for working with the cloned repository
    my $future = $git->run('status', '-s');
    my @stdout = $future->stdout;

=head1 SEE ALSO

L<Git::Repository>, L<Git::Wrapper>, L<Future>, L<Git::Version::Compare>

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report any bugs or feature requests through the issue tracker
at L<https://github.com/skirmess/Git-Background/issues>.
You will be notified automatically of any progress on your issue.

=head2 Source Code

This is open source software. The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/skirmess/Git-Background>

  git clone https://github.com/skirmess/Git-Background.git

=head1 AUTHOR

Sven Kirmess <sven.kirmess@kzone.ch>

=cut
